### DISCUSSION TITLE
Generic->Sketch_with_tft_setup  not found

### DISCUSSION THREAD

**Discussion by griffinmt** [Bodmer/TFT_eSPI#2189] [Q&A - General]
I currently have the most recent copy of your library (2.4.79) installed with Arduino IDE, but cannot find the mentioned sketch. Like many folks, I really want to do this. Can you help?

**Comment by Bodmer** [comment]
The setup  in sketch approach is now deprecated due to the number of issues it can cause in certain circumstances.

<del>It is here:
https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Generic/Sketch_with_tft_setup
 this only works on ESP32 and may cause problems with other libraries.</del>

**Reply by newdood** [reply]
Hi,
Unfortunately, the link returns an error for me.
is it possible to obtain this sketch?
Thanks ....

**Reply by Bodmer** [reply]
The sketch has recently been removed (see ReadMe) as the approach caused obscure problems with other libraries that are difficult to diagnose and fix.

**Reply by newdood** [reply]
Thank you for answer.
Is there another way to directly incorporate all information in the main sketch?
Is there a small tutorial or explanation that I could follow?

**Reply by Bodmer** [reply]
There seems to be no reliable approach with the normal Arduino IDE library folder use due to the way the compilation is handled (library has no visibility #defines in sketch).

Options:
1. Use PlatformIO IDE for Arduino sketches, add setup to project platformio.ini as detailed [here](https://github.com/Bodmer/TFT_eSPI/blob/master/docs/PlatformIO/Configuring%20options.txt).
2. Do not put the TFT_eSPI library in the library folder but include a copy with the correct setup defined in each sketch. I have not tried this but since the library is in the same folder as the sketch it should have visibility of the #defined in the sketch and the setup file. Clearly this creates a large sketch file size but these days that is not usually a problem. Advantage is you have a complete configuration controlled sketch so different sketches can contain a different version of the library. (Caveat I have not tried this!)
3. For ESP32, build under the ESP-IDF with the Arduino wrapper. Not something I have done but the library includes ESP-IDF menu config files provdided by a user.

If you figure out another viable approach then do post the solution!

**Reply by mcbp223** [reply]
> The setup in sketch approach is now deprecated due to the number of issues it can cause in certain circumstances.
> 
> ~It is here: https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Generic/Sketch_with_tft_setup this only works on ESP32 and may cause problems with other libraries.~

~~Hi, can you try with `compiler.cpp.extra_flags` and `compiler.c.extra_flags` instead of `build.extra_flags` in `platform.local.txt`? When using `build.extra_flags` the arduino builder appends a `@file` directive to the gcc command line right after the -I directive, restricting the search to the file in the cache folder, instead of the one in the source folder~~. EDIT: ~~it seems there's another side-effect of~~ adding '@' to the gcc command line, it's not directing it where to search for include files. First use of `@file` in this context was in Dec. 2017 in commit [Added recipe.hooks feature to override gcc options](https://github.com/stm32duino/Arduino_Core_STM32/commit/130c5ab3ba52d282a86cba2774481da9222b2484) for stm32.

This feature is too useful when dealing with different display configurations; even if it was impeded by the cache mechanism in IDE 2.0, with a warning when the ~~ARDUINO~~ IDE version detected is above 2 maybe it could stay? I'm not sure of which define it is though, with IDE 1.8.19 there's `-DARDUINO=10819`, and with 2.0.3 I see `-DARDUINO=10607`.

EDIT: I can't reproduce the sketch caching issue with either IDE, whenever I change something in tft_setup.h it gets copied into the TEMP sketch folder, and also TFT_eSPI.cpp gets recompiled. I'd like to do more tests on this if I could duplicate the conditions where the IDE skips recompilation of the library.

**Reply by newdood** [reply]
Thank you.
